### ISSUE TITLE
Since ver. 3.5 Conservative Caching removes plugin's JavaScript

### ISSUE THREAD

**Issue by vagari** [joomla/joomla-cms#11203]
#### Steps to reproduce the issue
1. Install Joomla
2. Create Contact in admin
3. Add "Contact Us" page menu item, single contact, that references created contact
4. Install betweenbrain/Custom-Contact-Form-Plugin - Amphenol Sensors branch (this is what's in production/tested)
5. Turn on Conservative Caching 
6. Visit Contact Us page on front end

#### Expected result
Small Javascript file is included in the header. If that file is updated to point at the correct form fields (the version on GitHub has generic #country & #state IDs) it will populate the state drop down when either United States or Canada is selected int he country drop down.

#### Actual result
With Conservative Caching enabled the Javascript is loaded the first time the page is accessed and it disappears on subsequent visits. At least until the cache expires or is cleared, then it appears a single time again. 

#### System information (as much as possible)
Linux
Database Version	5.5.50
PHP Version	5.6.23
Web Server	Apache
WebServer to PHP Interface	apache2handler
Joomla! Version	Joomla! 3.6.0 Stable 

#### Additional comments
Plugin worked correctly until 3.5 was installed.

The plugin can be found on [GitHub][1]. The Amphenol Sensors branch was written for me, developer hasn't worked with Joomla for a while so didn't know of a solution. 

Currently I have Conservative Caching disabled. And we're using the Plugin - System - Page Cache based on a workaround discussed on [Stack Overflow][2].

I've also tried some variations on when/how the Javascript file is included without success. Things like using `JHTML::script` instead of `->addScript`. And other execution times besides `onBeforeRender`. 

To rule out something else on the production site I sucessfully reproduced the issue on a clean install of Joomla 3.6. 

[1]:https://github.com/betweenbrain/Custom-Contact-Form-Plugin/tree/amphenol-sensors
[2]:https://stackoverflow.com/questions/38312055/joomla-3-5-cache-breaking-javascript-inclusion-for-contact-form-plugin

**Comment by csthomas** [comment]
Do you move:
```php
$this->doc->addScript(JURI::base(true) . '/media/plg_customcontactform/js/form-actions.min.js');
```
to method onContentPrepareForm

**Comment by vagari** [comment]
Hi @csthomas, Are you asking if I attempted that? I did try that but it migrated the Javascript to the beginning of the "chain," before anything else, including jQuery, was loaded. Which fixed the cache issue, but broke the actual Javascript.

**Comment by csthomas** [comment]
Yes, I updated my question.

Then add jquery before that line.

IMHO this is not joomla bug.
You should go to forum.joomla.org if you can not resolve it yourself. 

If you use something that requires jquery then you have to load jquery before (ex. line before).
This will change "chain".

**Comment by vagari** [comment]
Fair enough. I understand the concept. Migrating the script to `onContentPrepareForm` from `onBeforeRender` and including the `JHtml::_('jquery.framework')` solves the caching issue. (Though it does still places it above a bunch of other JavaScript so it might not work for others). And it may not be a bug in the proper sense, but considering many plugins use `onBeforeRender`, (and I've seen it in plenty of tutorials) I'm still surprised at the outcome. Is there an actual reason (apparently undocumented) that caching does not work in this scenario. 

TL:DR I'd still consider this a workaround as opposed to a fix at this juncture.

**Comment by vagari** [comment]
Oh, and to reiterate from my original post, the way the plugin was written functioned correctly before the release of Joomla 3.5.

**Comment by csthomas** [comment]
Try something direct instead ```JHtml::_('jquery.framework')```
$doc->addScript...

**Comment by vagari** [comment]
Are you saying, instead of referencing the jQuery framework via JHtml, use an explicit addScript to add jQuery?

**Comment by csthomas** [comment]
yes, there may be bug with caching but I can not confirm now.

**Comment by weeblr** [comment]
Looking into this #jab17<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11203">issues.joomla.org/tracker/joomla-cms/11203</a>.</sub>

**Comment by andrepereiradasilva** [comment]
further info that may be related: see also https://github.com/joomla/joomla-cms/issues/16375#issuecomment-305864460 reggarding modules and conservative caching

**Comment by weeblr** [comment]
Hi all,

This is happening because when Caching is set to "Conservative", the component output is fetched from the cache, meaning any code from view.html.php is not executed. This includes the:

`JPluginHelper::importPlugin('content');`

instructions. This plugin is a content plugin, and except on 1st page load, it's simply not loaded. Therefore it onBeforeRender handler is not registered and not executed. 

I would think a content plugin should be allowed to insert js/css into the page head section and so content plugins should also be loaded when conservative caching is enabled. Not been able to pinpoint when this was changed around 3.5 release.

/cc @csthomas @mbabker @nibra <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11203">issues.joomla.org/tracker/joomla-cms/11203</a>.</sub>

**Comment by mbabker** [comment]
Maybe we have the app ensure content plugins are imported before `onAfterDispatch` is dispatched?

**Comment by weeblr** [comment]
Actually it's enough to move the plugin loading to the content controller vs in each individual per view view.html.php, as it is today. I have confirmed that.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11203">issues.joomla.org/tracker/joomla-cms/11203</a>.</sub>

**Comment by nibra** [comment]
Added information to [documentation](https://docs.joomla.org/J3.x:Developing_an_MVC_Component/Adding_a_view_to_the_site_part#view.html.php), closing this issue.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11203">issues.joomla.org/tracker/joomla-cms/11203</a>.</sub>

**Comment by andrepereiradasilva** [comment]
@nibra so if i understand properly this one https://github.com/joomla/joomla-cms/issues/16375 is the same issuie and can also be closed with the same motive?

**Comment by nibra** [comment]
Seems to be the same issue, yes.
